src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/BodySubstitutionParser.cs

```diff
@@ -1,174 +1,145 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
-using System.Linq;
+using System.Reflection.Metadata;
+using Internal.TypeSystem;
 using System.Xml.XPath;
-using ILLink.Shared;
-using Mono.Cecil;
+using System.Globalization;
 
-namespace Mono.Linker.Steps
+namespace ILCompiler
 {
-    public class BodySubstitutionParser : ProcessLinkerXmlBase
+    internal class SubstitutionsParser : ProcessLinkerXmlBase
     {
-        SubstitutionInfo? _substitutionInfo;
-
-        public BodySubstitutionParser(LinkContext context, Stream documentStream, string xmlDocumentLocation)
-            : base(context, documentStream, xmlDocumentLocation)
-        {
-        }
+        private readonly Dictionary<MethodDesc, BodySubstitution> _methodSubstitutions;
+        private readonly Dictionary<FieldDesc, object> _fieldSubstitutions;
 
-        public BodySubstitutionParser(LinkContext context, Stream documentStream, EmbeddedResource resource, AssemblyDefinition resourceAssembly, string xmlDocumentLocation = "")
-            : base(context, documentStream, resource, resourceAssembly, xmlDocumentLocation)
+        private SubstitutionsParser(TypeSystemContext context, UnmanagedMemoryStream documentStream, ManifestResource resource, ModuleDesc resourceAssembly, string xmlDocumentLocation, IReadOnlyDictionary<string, bool> featureSwitchValues)
```

Comment:
Could you keep `documentStream` as `Stream` instead of making it `UnmanagedMemoryStream`? If we want to support the equivalent of `--substitutions` command line option, we'll probably not open the file as a memory mapped file.